### PR TITLE
docs/op-guide: create GC document

### PR DESCRIPTION
**Pull request by CaitinChen**
@MyonKeminta  @lilin90 PTAL
via https://github.com/pingcap/docs-cn/pull/679

### CODE REVIEW

File: op-guide/gc.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+## Implementation details
+
+The GC implementation process is complex. Clearing the data that is not used any more should be on the premise that data consistency is guaranteed. The process of doing GC is as below:
+
+### 1. Resolve locks
+
+The TiDB transaction is based on Google Percolator. The transaction committing is a two-phase committing process. When the first phase is finished, all the related keys are locked. Among these locks, one is the primary lock and the others are secondary locks which contain a pointer of the primary locks; in the secondary phase, the key with the primary lock gets a write record and its lock is removed. The write record indicates the write or delete operation in the history or the transactional rollback record of this key. Replacing the primary lock by which write record indicates whether the corresponding transaction is committed successfully. Then all the secondary locks are replaced successively. If the threads to replace the secondary locks fail, these locks are retained. During GC, locks whose timestamp is before the safe point will be replaced by the corresponding write record based on the transaction committing status.
```
Comment (lilin90) [inline op-guide/gc.md]: - "The TiDB transaction is based on Google Percolator. The transaction committing is a two-phase committing process." -> "The TiDB transaction model is inspired by Google's Percolator. It's mainly a two-phase commit protocol with some practical optimizations."

    Ref: https://pingcap.com/blog/2016-10-17-how-we-build-tidb/#12

- "Replacing the primary lock by which" -> "Replacing the primary lock **with** which"

- "locks whose timestamp is before the safe point will be" -> "**the lock** whose timestamp is before the safe point **is**"

File: op-guide/gc.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+This step is necessary, because you are not informed whether this transaction is successful if GC has cleared the write record of the primary lock, thus data consistency cannot be guaranteed.
```
Comment (lilin90) [inline op-guide/gc.md]: Change to:

> **Note:** This is a required step. Once GC has cleared the write record of the primary lock, you can never know whether this transaction is successful or not. As a result, data consistency cannot be guaranteed.

File: op-guide/gc.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+### 2. Delete ranges
+
+The `DeleteRanges` operation is usually necessary after the operation such as `drop table`. It is used to delete a range which may be very large. If the `use_delete_range` option of TiKV is not enabled, TiKV deletes the keys in the range.
```
Comment (lilin90) [inline op-guide/gc.md]: "The `DeleteRanges` operation is usually necessary after the operation such as `drop table`. It is used to delete a range which may be very large." -> "`DeleteRanges` is usually executed after operations like `drop table`, used to delete a range which might be very large."

**Comment by CaitinChen** [comment]
@MyonKeminta @lilin90 PTAL Thanks

File: op-guide/gc.md [hunk 4]
```diff
@@ -36,42 +36,48 @@ mysql> select VARIABLE_NAME, VARIABLE_VALUE from mysql.tidb;
...
 
-> **Note**: When you set `tikv_gc_life_time` to a large number (like days or even months) in a data updated frequently scenario, some problems as follows may occur: 
+    > **Note**: When you set `tikv_gc_life_time` to a large number (like days or even months) in a data updated frequently scenario, some problems as follows may occur: 
 
-    - The more versions of the data, the more disk storage space is occupied.
-    - A large number of history versions might slow down the query. They may affect range queries like `select count(*) from t`.
-    - If `tikv_gc_life_time` is suddenly turned to a smaller value during operation, a great deal of old data may be deleted in a short time, causing I/O pressure.
+        - The more versions of the data, the more disk storage space is occupied.
+        - A large number of history versions might slow down the query. They may affect range queries like `select count(*) from t`.
+        - If `tikv_gc_life_time` is suddenly turned to a smaller value during operation, a great deal of old data may be deleted in a short time, causing I/O pressure.
```
Comment (lilin90) [inline op-guide/gc.md]: Please check the format here. It displays as a code block.

File: op-guide/gc.md [hunk 5]
```diff
@@ -36,42 +36,48 @@ mysql> select VARIABLE_NAME, VARIABLE_VALUE from mysql.tidb;
...
 
-This step is necessary, because you are not informed whether this transaction is successful if GC has cleared the write record of the primary lock, thus data consistency cannot be guaranteed.
+**Note**: This is a required step. Once GC has cleared the write record of the primary lock, you can never know whether this transaction is successful or not. As a result, data consistency cannot be guaranteed.
```
Comment (lilin90) [inline op-guide/gc.md]: Use the "Note" format.

**Comment by CaitinChen** [comment]
@lilin90 @MyonKeminta PTAL gain

File: op-guide/gc.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+> **Note**: This is a required step. Once GC has cleared the write record of the primary lock, you can never know whether this transaction is successful or not. As a result, data consistency cannot be guaranteed.
+
+### 2. Delete ranges
+
+`DeleteRanges` is usually executed after operations like `drop table`, used to delete a range which might be very large. If the `use_delete_range` option of TiKV is not enabled, TiKV deletes the keys in the range.
+
+### 3. Do GC
+
+Clear the data before the safe point of each key and the write record. 
+> **Note**: if the last record in all the write records of `Put` and `Delete` types before the safe point is `Put`, this record and its data cannot be deleted directly. Otherwise, you cannot successfully perform the read operation whose timestamp is after the safe point and before the next key version.
```
Comment (lilin90) [inline op-guide/gc.md]: if -> If

File: op-guide/gc.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
+---
+title: TiDB Garbage Collection (GC)
+category: advanced
+---
+
+# TiDB Garbage Collection (GC)
+
+TiDB uses MVCC to control concurrency. When you update or delete data, the data is not deleted immediately but is saved for a period during which it can be read. Thus the write operation and the read operation are not mutually exclusive and it is possible to read the previous data.
```
Comment (MyonKeminta) [inline op-guide/gc.md]: Maybe it's better to say "the **original** data is not deleted immediately but is **kept** for a while ....."

Comment (MyonKeminta) [reply op-guide/gc.md]: "It is possible to read the **history versions** of the data". See op-guide/history-read.md

File: op-guide/gc.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+The data versions whose duration exceeds a specific time and that are not used any more will be cleared, otherwise they will occupy the disk space, affecting the system performance. TiDB uses Garbage Collection (GC) to clear the obsolete data. 
+
+## Working mechanism
+
+GC runs periodically on TiDB. When a TiDB server is started, a `gc_worker` is enabled in the background. In each TiDB cluster, one `gc_worker` is elected to be the leader which is used to maintain the GC status and send GC commands to all the TiKV Region leaders.
```
Comment (MyonKeminta) [inline op-guide/gc.md]: A gc worker is started in the  background, not enabled. These text are moved from history-read.md, around line 34.

File: op-guide/gc.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+The data versions whose duration exceeds a specific time and that are not used any more will be cleared, otherwise they will occupy the disk space, affecting the system performance. TiDB uses Garbage Collection (GC) to clear the obsolete data. 
+
+## Working mechanism
+
+GC runs periodically on TiDB. When a TiDB server is started, a `gc_worker` is enabled in the background. In each TiDB cluster, one `gc_worker` is elected to be the leader which is used to maintain the GC status and send GC commands to all the TiKV Region leaders. 
+
+## Configuration and monitor
+
+The GC configuration and operational status are recorded in the `mysql.tidb` system table, which can be monitored and configured using the following SQL statement:
```
Comment (MyonKeminta) [inline op-guide/gc.md]: Maybe not "following"...

File: op-guide/gc.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+| tikv_gc_run_interval  | 10m0s                                                                                          |
+| tikv_gc_life_time     | 10m0s                                                                                          |
+| tikv_gc_last_run_time | 20180418-10:59:30 +0800 CST                                                                    |
+| tikv_gc_safe_point    | 20180418-10:58:30 +0800 CST                                                                    |
+| tikv_gc_concurrency   | 1                                                                                              |
++-----------------------+------------------------------------------------------------------------------------------------+
+10 rows in set (0.02 sec)
+```
+
+In the table above, `tikv_gc_run_interval`, `tikv_gc_life_time` and `tikv_gc_concurrency` can be configured manually. Other variables with the `tikv_gc prefix` prefix record the current status, which are automatically updated by TiDB. Do not modify these variables.
```
Comment (MyonKeminta) [inline op-guide/gc.md]: `tikv_gc` prefix

**Comment by CaitinChen** [comment]
@MyonKeminta  @lilin90  PTAL again

File: op-guide/gc.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+> **Note**: This is a required step. Once GC has cleared the write record of the primary lock, you can never know whether this transaction is successful or not. As a result, data consistency cannot be guaranteed.
+
+### 2. Delete ranges
+
+`DeleteRanges` is usually executed after operations like `drop table`, used to delete a range which might be very large. If the `use_delete_range` option of TiKV is not enabled, TiKV deletes the keys in the range.
+
+### 3. Do GC
+
+Clear the data before the safe point of each key and the write record. 
+> **Note**: If the last record in all the write records of `Put` and `Delete` types before the safe point is `Put`, this record and its data cannot be deleted directly. Otherwise, you cannot successfully perform the read operation whose timestamp is after the safe point and before the next key version.
```
Comment (MyonKeminta) [inline op-guide/gc.md]: ... after the safe point and before the **next version of the key**

**Comment by MyonKeminta** [comment]
Rest LGTM

**Comment by CaitinChen** [comment]
@MyonKeminta Thanks

**Comment by lilin90** [comment]
@MyonKeminta Thanks a lot for your comments. 😀

**Pull request merged**